core/src/main/java/com/linecorp/armeria/common/ArmeriaOptionsProvider.java

```diff
@@ -0,0 +1,424 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.net.InetAddress;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.common.util.TransportType;
+import com.linecorp.armeria.server.TransientServiceOption;
+
+/**
+ * A Java SPI (Service Provider Interface) for the {@link Flags} value.
+ * Value provides from Java SPI will be ignored if there corresponding valid JVM option.
+ */
+@UnstableApi
+public interface ArmeriaOptionsProvider {
```

Comment:
- How about just `FlagsProvider`?
- How about allowing all methods to return `null`, so that a custom provider implementation can choose to return `null` when it wants to use the default value? e.g.
  ```
  @Nullable
  default Boolean verboseSocketExceptions() {
      return null;
  }
  ```

Reply:
> How about just FlagsProvider?

That sounds good to me.

> How about allowing all methods to return null, so that a custom provider implementation can choose to return null when it wants to use the default value? Once we allow returning null in the provider, we can move this back into `Flags`.

I agreed. Thanks for the suggestion🙏.